File: 0518-coin-change-ii/0518-coin-change-ii.java
class Solution {

    int help(int idx , int amount, int[] coins, int[][]dp){
    if(amount == 0) return 1;
       if(idx == coins.length){
        if(amount == 0) return 1;
        return 0;
       }

       if(dp[idx][amount] != -1) return dp[idx][amount];

       int pick=0 , npick = 0;

       if(coins[idx] <= amount) pick = help(idx,amount-coins[idx],coins,dp);
       npick = help(idx+1,amount,coins,dp);

       return dp[idx][amount] = pick + npick;
    }
    public int change(int amount, int[] coins) {
        //if(amount == 0) return 0;
         int[][] dp = new int[coins.length][amount+1];
        for(int[] a: dp){
            Arrays.fill(a,-1);
        }

        return help(0,amount,coins,dp);
    }
}

File: 0518-coin-change-ii/0518-coin-change-ii.cpp
class Solution {
public:
int helper(int idx,int amt, vector<int>& arr,vector<vector<int>>&dp){

    //if(amt ==0) return 1;
    if(amt ==0) return 1;
    if(idx == arr.size()-1){ 
        if(amt%arr[idx] ==0) return 1;
        return 0;
        }
 if(dp[idx][amt] != -1) return dp[idx][amt];
    int pick = 0;
    if(arr[idx]<= amt){
        pick = helper(idx,amt-arr[idx],arr,dp);
    }

    int nopick = helper(idx+1,amt,arr,dp);

    return dp[idx][amt] = pick+nopick;

}
    int change(int amount, vector<int>& coins) {
         vector<vector<int>>dp(coins.size()+1,vector<int>(amount+1,-1));

         return helper(0,amount,coins,dp);
    }
};